salt/states/zfs.py

```diff
@@ -341,104 +394,194 @@ def filesystem_present(name, create_parent=False, properties=None, cloned_from=N
+        elif ret['result']:
+            ret['comment'] = '{0} {1} is uptodate'.format(dataset_type, name)
         else:
-            ret['comment'] = 'filesystem {0} is up to date'.format(name)
-    else:  # create filesystem
-        result = {name: 'created'}
-        if not __opts__['test']:
-            if not cloned_from:
-                result = __salt__['zfs.create'](name, **{'create_parent': create_parent, 'properties': properties})
-            else:
-                result = __salt__['zfs.clone'](cloned_from, name, **{'create_parent': create_parent, 'properties': properties})
+            ret['comment'] = '{0} {1} failed to be updated'.format(dataset_type, name)
 
-        ret['result'] = name in result
-        if ret['result']:
-            ret['result'] = result[name] == 'created' or result[name].startswith('cloned')
+    ## NOTE: create or clone the dataset
+    else:
+        mod_res_action = 'cloned' if cloned_from else 'created'
+        if __opts__['test']:
+            ## NOTE: pretend to create/clone
+            mod_res = OrderedDict([
+                (mod_res_action, True),
+            ])
+        elif cloned_from:
+            ## NOTE: add volsize to properties
+            if volume_size:
+                properties['volsize'] = volume_size
+
+            ## NOTE: clone the dataset
+            mod_res = __salt__['zfs.clone'](cloned_from, name, **{
+                'create_parent': create_parent,
+                'properties': properties,
+            })
+        else:
+            ## NOTE: create the dataset
+            mod_res = __salt__['zfs.create'](name, **{
+                'create_parent': create_parent,
+                'properties': properties,
+                'volume_size': volume_size,
+                'sparse': sparse,
+            })
+
+        ret['result'] = mod_res[mod_res_action]
         if ret['result']:
-            ret['changes'][name] = properties if len(properties) > 0 else result[name]
-            ret['comment'] = 'filesystem {0} was created'.format(name)
+            ret['changes'][name] = mod_res_action
+            if properties:
+                ret['changes'][name] = properties
+            ret['comment'] = '{0} {1} was {2}'.format(
+                dataset_type,
+                name,
+                mod_res_action,
+            )
         else:
-            ret['comment'] = 'failed to create filesystem {0}'.format(name)
-            if name in result:
-                ret['comment'] = result[name]
+            ret['comment'] = 'failed to {0} {1} {2}'.format(
+                mod_res_action[:-1],
+                dataset_type,
+                name,
+            )
+            if 'error' in mod_res:
+                ret['comment'] = mod_res['error']
+
     return ret
 
 
+def filesystem_present(name, create_parent=False, properties=None, cloned_from=None):
+    '''
+    ensure filesystem exists and has properties set
+
```

Comment:
This function could use a `.. versionadded:: Fluorine`

Reply:
Why? This fucntions has been around since way before Flourine?

Reply:
Ah, sorry. The way the diff came through made it look like this was a new function.